File: agents/ten_packages/extension/elevenlabs_tts/elevenlabs_tts_extension.go
/**
 *
 * Agora Real Time Engagement
 * Created by XinHui Li in 2024-07.
 * Copyright (c) 2024 Agora IO. All rights reserved.
 *
 */
// Note that this is just an example extension written in the GO programming
// language, so the package name does not equal to the containing directory
// name. However, it is not common in Go.
package extension

import (
	"fmt"
	"io"
	"sync"
	"sync/atomic"
	"time"

	"ten_framework/ten"
)

const (
	cmdInFlush                 = "flush"
	cmdOutFlush                = "flush"
	dataInTextDataPropertyText = "text"

	propertyApiKey                   = "api_key"                    // Required
	propertyModelId                  = "model_id"                   // Optional
	propertyOptimizeStreamingLatency = "optimize_streaming_latency" // Optional
	propertyRequestTimeoutSeconds    = "request_timeout_seconds"    // Optional
	propertySimilarityBoost          = "similarity_boost"           // Optional
	propertySpeakerBoost             = "speaker_boost"              // Optional
	propertyStability                = "stability"                  // Optional
	propertyStyle                    = "style"                      // Optional
	propertyVoiceId                  = "voice_id"                   // Optional
)

const (
	textChanMax = 1024
)

var (
	outdateTs atomic.Int64
	textChan  chan *message
	wg        sync.WaitGroup
)

type elevenlabsTTSExtension struct {
	ten.DefaultExtension
	elevenlabsTTS *elevenlabsTTS
}

type message struct {
	text       string
	receivedTs int64
}

func newElevenlabsTTSExtension(name string) ten.Extension {
	return &elevenlabsTTSExtension{}
}

// OnStart will be called when the extension is starting,
// properies can be read here to initialize and start the extension.
// current supported properties:
//   - api_key (required)
//   - model_id
//   - optimize_streaming_latency
//   - request_timeout_seconds
//   - similarity_boost
//   - speaker_boost
//   - stability
//   - style
//   - voice_id
func (e *elevenlabsTTSExtension) OnStart(ten ten.TenEnv) {
	ten.LogInfo("OnStart")

	// prepare configuration
	elevenlabsTTSConfig := defaultElevenlabsTTSConfig()

	if apiKey, err := ten.GetPropertyString(propertyApiKey); err != nil {
		ten.LogError(fmt.Sprintf("GetProperty required %s failed, err: %v", propertyApiKey, err))
		return
	} else {
		elevenlabsTTSConfig.ApiKey = apiKey
	}

	if modelId, err := ten.GetPropertyString(propertyModelId); err != nil {
		ten.LogWarn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyModelId, err))
	} else {
		if len(modelId) > 0 {
			elevenlabsTTSConfig.ModelId = modelId
		}
	}

	if optimizeStreamingLatency, err := ten.GetPropertyInt64(propertyOptimizeStreamingLatency); err != nil {
		ten.LogWarn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyOptimizeStreamingLatency, err))
	} else {
		if optimizeStreamingLatency > 0 {
			elevenlabsTTSConfig.OptimizeStreamingLatency = int(optimizeStreamingLatency)
		}
	}

	if requestTimeoutSeconds, err := ten.GetPropertyInt64(propertyRequestTimeoutSeconds); err != nil {
		ten.LogWarn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyRequestTimeoutSeconds, err))
	} else {
		if requestTimeoutSeconds > 0 {
			elevenlabsTTSConfig.RequestTimeoutSeconds = int(requestTimeoutSeconds)
		}
	}

	if similarityBoost, err := ten.GetPropertyFloat64(propertySimilarityBoost); err != nil {
		ten.LogWarn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertySimilarityBoost, err))
	} else {
		elevenlabsTTSConfig.SimilarityBoost = float32(similarityBoost)
	}

	if speakerBoost, err := ten.GetPropertyBool(propertySpeakerBoost); err != nil {
		ten.LogWarn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertySpeakerBoost, err))
	} else {
		elevenlabsTTSConfig.SpeakerBoost = speakerBoost
	}

	if stability, err := ten.GetPropertyFloat64(propertyStability); err != nil {
		ten.LogWarn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyStability, err))
	} else {
		elevenlabsTTSConfig.Stability = float32(stability)
	}

	if style, err := ten.GetPropertyFloat64(propertyStyle); err != nil {
		ten.LogWarn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyStyle, err))
	} else {
		elevenlabsTTSConfig.Style = float32(style)
	}

	if voiceId, err := ten.GetPropertyString(propertyVoiceId); err != nil {
		ten.LogWarn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyVoiceId, err))
	} else {
		if len(voiceId) > 0 {
			elevenlabsTTSConfig.VoiceId = voiceId
		}
	}

	// create elevenlabsTTS instance
	elevenlabsTTS, err := newElevenlabsTTS(elevenlabsTTSConfig)
	if err != nil {
		ten.LogError(fmt.Sprintf("newElevenlabsTTS failed, err: %v", err))
		return
	}

	ten.LogInfo(fmt.Sprintf("newElevenlabsTTS succeed with ModelId: %s, VoiceId: %s",
		elevenlabsTTSConfig.ModelId, elevenlabsTTSConfig.VoiceId))

	// set elevenlabsTTS instance
	e.elevenlabsTTS = elevenlabsTTS

	// create pcm instance
	pcm := newPcm(defaultPcmConfig())
	pcmFrameSize := pcm.getPcmFrameSize()

	// init chan
	textChan = make(chan *message, textChanMax)

	go func() {
		ten.LogInfo("process textChan")

		for msg := range textChan {
			if msg.receivedTs < outdateTs.Load() { // Check whether to interrupt
				ten.LogInfo(fmt.Sprintf("textChan interrupt and flushing for input text: [%s], receivedTs: %d, outdateTs: %d",
					msg.text, msg.receivedTs, outdateTs.Load()))
				continue
			}

			wg.Add(1)
			ten.LogInfo(fmt.Sprintf("textChan text: [%s]", msg.text))

			r, w := io.Pipe()
			startTime := time.Now()

			go func() {
				defer wg.Done()
				defer w.Close()

				ten.LogInfo(fmt.Sprintf("textToSpeechStream text: [%s]", msg.text))

				err = e.elevenlabsTTS.textToSpeechStream(w, msg.text)
				if err != nil {
					ten.LogError(fmt.Sprintf("textToSpeechStream failed, err: %v", err))
					return
				}
			}()

			ten.LogInfo(fmt.Sprintf("read pcm stream, text:[%s], pcmFrameSize:%d", msg.text, pcmFrameSize))

			var (
				firstFrameLatency int64
				n                 int
				pcmFrameRead      int
				readBytes         int
				sentFrames        int
			)
			buf := pcm.newBuf()

			// read pcm stream
			for {
				if msg.receivedTs < outdateTs.Load() { // Check whether to interrupt
					ten.LogInfo(fmt.Sprintf("read pcm stream interrupt and flushing for input text: [%s], receivedTs: %d, outdateTs: %d",
						msg.text, msg.receivedTs, outdateTs.Load()))
					break
				}

				n, err = r.Read(buf[pcmFrameRead:])
				readBytes += n
				pcmFrameRead += n

				if err != nil {
					if err == io.EOF {
						ten.LogInfo("read pcm stream EOF")
						break
					}

					ten.LogError(fmt.Sprintf("read pcm stream failed, err: %v", err))
					break
				}

				if pcmFrameRead != pcmFrameSize {
					ten.LogDebug(fmt.Sprintf("the number of bytes read is [%d] inconsistent with pcm frame size", pcmFrameRead))
					continue
				}

				pcm.send(ten, buf)
				// clear buf
				buf = pcm.newBuf()
				pcmFrameRead = 0
				sentFrames++

				if firstFrameLatency == 0 {
					firstFrameLatency = time.Since(startTime).Milliseconds()
					ten.LogInfo(fmt.Sprintf("first frame available for text: [%s], receivedTs: %d, firstFrameLatency: %dms", msg.text, msg.receivedTs, firstFrameLatency))
				}

				ten.LogDebug(fmt.Sprintf("sending pcm data, text: [%s]", msg.text))
			}

			if pcmFrameRead > 0 {
				pcm.send(ten, buf)
				sentFrames++
				ten.LogInfo(fmt.Sprintf("sending pcm remain data, text: [%s], pcmFrameRead: %d", msg.text, pcmFrameRead))
			}

			r.Close()
			ten.LogInfo(fmt.Sprintf("send pcm data finished, text: [%s], receivedTs: %d, readBytes: %d, sentFrames: %d, firstFrameLatency: %dms, finishLatency: %dms",
				msg.text, msg.receivedTs, readBytes, sentFrames, firstFrameLatency, time.Since(startTime).Milliseconds()))
		}
	}()

	ten.OnStartDone()
}

// OnCmd receives cmd from ten graph.
// current supported cmd:
//   - name: flush
//     example:
//     {"name": "flush"}
func (e *elevenlabsTTSExtension) OnCmd(
	tenEnv ten.TenEnv,
	cmd ten.Cmd,
) {
	cmdName, err := cmd.GetName()
	if err != nil {
		tenEnv.LogError(fmt.Sprintf("OnCmd get name failed, err: %v", err))
		cmdResult, _ := ten.NewCmdResult(ten.StatusCodeError)
		tenEnv.ReturnResult(cmdResult, cmd, nil)
		return
	}

	tenEnv.LogInfo(fmt.Sprintf("OnCmd %s", cmdInFlush))

	switch cmdName {
	case cmdInFlush:
		outdateTs.Store(time.Now().UnixMicro())

		// send out
		outCmd, err := ten.NewCmd(cmdOutFlush)
		if err != nil {
			tenEnv.LogError(fmt.Sprintf("new cmd %s failed, err: %v", cmdOutFlush, err))
			cmdResult, _ := ten.NewCmdResult(ten.StatusCodeError)
			tenEnv.ReturnResult(cmdResult, cmd, nil)
			return
		}

		if err := tenEnv.SendCmd(outCmd, nil); err != nil {
			tenEnv.LogError(fmt.Sprintf("send cmd %s failed, err: %v", cmdOutFlush, err))
			cmdResult, _ := ten.NewCmdResult(ten.StatusCodeError)
			tenEnv.ReturnResult(cmdResult, cmd, nil)
			return
		} else {
			tenEnv.LogInfo(fmt.Sprintf("cmd %s sent", cmdOutFlush))
		}
	}

	cmdResult, _ := ten.NewCmdResult(ten.StatusCodeOk)
	tenEnv.ReturnResult(cmdResult, cmd, nil)
}

// OnData receives data from ten graph.
// current supported data:
//   - name: text_data
//     example:
//     {name: text_data, properties: {text: "hello"}
func (e *elevenlabsTTSExtension) OnData(
	tenEnv ten.TenEnv,
	data ten.Data,
) {
	text, err := data.GetPropertyString(dataInTextDataPropertyText)
	if err != nil {
		tenEnv.LogWarn(fmt.Sprintf("OnData GetProperty %s failed, err: %v", dataInTextDataPropertyText, err))
		return
	}

	if len(text) == 0 {
		tenEnv.LogDebug("OnData text is empty, ignored")
		return
	}

	tenEnv.LogInfo(fmt.Sprintf("OnData input text: [%s]", text))

	go func() {
		textChan <- &message{text: text, receivedTs: time.Now().UnixMicro()}
	}()
}

func init() {
	// Register addon
	ten.RegisterAddonAsExtension(
		"elevenlabs_tts",
		ten.NewDefaultExtensionAddon(newElevenlabsTTSExtension),
	)
}

File: agents/ten_packages/extension/elevenlabs_tts_python/elevenlabs_tts_extension.py
#
#
# Agora Real Time Engagement
# Created by XinHui Li in 2024-07.
# Copyright (c) 2024 Agora IO. All rights reserved.
#
#

import queue
import threading
import time

from ten import (
    Extension,
    TenEnv,
    Cmd,
    CmdResult,
    StatusCode,
    Data,
)
from .elevenlabs_tts import default_elevenlabs_tts_config, ElevenlabsTTS
from .pcm import PcmConfig, Pcm
from .log import logger

CMD_IN_FLUSH = "flush"
CMD_OUT_FLUSH = "flush"

DATA_IN_TEXT_DATA_PROPERTY_TEXT = "text"

PROPERTY_API_KEY = "api_key"  # Required
PROPERTY_MODEL_ID = "model_id"  # Optional
PROPERTY_OPTIMIZE_STREAMING_LATENCY = "optimize_streaming_latency"  # Optional
PROPERTY_REQUEST_TIMEOUT_SECONDS = "request_timeout_seconds"  # Optional
PROPERTY_SIMILARITY_BOOST = "similarity_boost"  # Optional
PROPERTY_SPEAKER_BOOST = "speaker_boost"  # Optional
PROPERTY_STABILITY = "stability"  # Optional
PROPERTY_STYLE = "style"  # Optional


class Message:
    def __init__(self, text: str, received_ts: int) -> None:
        self.text = text
        self.received_ts = received_ts


class ElevenlabsTTSExtension(Extension):
    def on_start(self, ten: TenEnv) -> None:
        logger.info("on_start")

        self.elevenlabs_tts = None
        self.outdate_ts = 0
        self.pcm = None
        self.pcm_frame_size = 0
        self.text_queue = queue.Queue(maxsize=1024)

        # prepare configuration
        elevenlabs_tts_config = default_elevenlabs_tts_config()

        try:
            elevenlabs_tts_config.api_key = ten.get_property_string(PROPERTY_API_KEY)
        except Exception as e:
            logger.warning(f"on_start get_property_string {PROPERTY_API_KEY} error: {e}")
            return

        try:
            model_id = ten.get_property_string(PROPERTY_MODEL_ID)
            if len(model_id) > 0:
                elevenlabs_tts_config.model_id = model_id
        except Exception as e:
            logger.warning(f"on_start get_property_string {PROPERTY_MODEL_ID} error: {e}")

        try:
            optimize_streaming_latency = ten.get_property_int(PROPERTY_OPTIMIZE_STREAMING_LATENCY)
            if optimize_streaming_latency > 0:
                elevenlabs_tts_config.optimize_streaming_latency = optimize_streaming_latency
        except Exception as e:
            logger.warning(f"on_start get_property_int {PROPERTY_OPTIMIZE_STREAMING_LATENCY} error: {e}")

        try:
            request_timeout_seconds = ten.get_property_int(PROPERTY_REQUEST_TIMEOUT_SECONDS)
            if request_timeout_seconds > 0:
                elevenlabs_tts_config.request_timeout_seconds = request_timeout_seconds
        except Exception as e:
            logger.warning(f"on_start get_property_int {PROPERTY_REQUEST_TIMEOUT_SECONDS} error: {e}")

        try:
            elevenlabs_tts_config.similarity_boost = ten.get_property_float(PROPERTY_SIMILARITY_BOOST)
        except Exception as e:
            logger.warning(f"on_start get_property_float {PROPERTY_SIMILARITY_BOOST} error: {e}")

        try:
            elevenlabs_tts_config.speaker_boost = ten.get_property_bool(PROPERTY_SPEAKER_BOOST)
        except Exception as e:
            logger.warning(f"on_start get_property_bool {PROPERTY_SPEAKER_BOOST} error: {e}")

        try:
            elevenlabs_tts_config.stability = ten.get_property_float(PROPERTY_STABILITY)
        except Exception as e:
            logger.warning(f"on_start get_property_float {PROPERTY_STABILITY} error: {e}")

        try:
            elevenlabs_tts_config.style = ten.get_property_float(PROPERTY_STYLE)
        except Exception as e:
            logger.warning(f"on_start get_property_float {PROPERTY_STYLE} error: {e}")

        # create elevenlabsTTS instance
        self.elevenlabs_tts = ElevenlabsTTS(elevenlabs_tts_config)

        logger.info(f"ElevenlabsTTS succeed with model_id: {self.elevenlabs_tts.config.model_id}, VoiceId: {self.elevenlabs_tts.config.voice_id}")

        # create pcm instance
        self.pcm = Pcm(PcmConfig())
        self.pcm_frame_size = self.pcm.get_pcm_frame_size()

        threading.Thread(target=self.process_text_queue, args=(ten,)).start()

        ten.on_start_done()

    def on_stop(self, ten: TenEnv) -> None:
        logger.info("on_stop")
        ten.on_stop_done()

    def on_cmd(self, ten: TenEnv, cmd: Cmd) -> None:
        """
        on_cmd receives cmd from ten graph.
        current supported cmd:
          - name: flush
            example:
            {"name": "flush"}
        """
        logger.info("on_cmd")
        cmd_name = cmd.get_name()

        logger.info(f"on_cmd [{cmd_name}]")

        if cmd_name is CMD_IN_FLUSH:
            self.outdate_ts = int(time.time() * 1000000)

            # send out
            out_cmd = Cmd.create(CMD_OUT_FLUSH)
            ten.send_cmd(out_cmd)

        cmd_result = CmdResult.create(StatusCode.OK)
        cmd_result.set_property_string("detail", "success")
        ten.return_result(cmd_result, cmd)

    def on_data(self, ten: TenEnv, data: Data) -> None:
        """
        on_data receives data from ten graph.
        current supported data:
          - name: text_data
            example:
            {name: text_data, properties: {text: "hello"}
        """
        logger.info("on_data")

        try:
            text = data.get_property_string(DATA_IN_TEXT_DATA_PROPERTY_TEXT)
        except Exception as e:
            logger.warning(f"on_data get_property_string {DATA_IN_TEXT_DATA_PROPERTY_TEXT} error: {e}")
            return

        if len(text) == 0:
            logger.debug("on_data text is empty, ignored")
            return

        logger.info(f"OnData input text: [{text}]")

        self.text_queue.put(Message(text, int(time.time() * 1000000)))

    def process_text_queue(self, ten: TenEnv):
        logger.info("process_text_queue")

        while True:
            msg = self.text_queue.get()
            logger.debug(f"process_text_queue, text: [{msg.text}]")

            if msg.received_ts < self.outdate_ts:
                logger.info(f"textChan interrupt and flushing for input text: [{msg.text}], received_ts: {msg.received_ts}, outdate_ts: {self.outdate_ts}")
                continue

            start_time = time.time()
            buf = self.pcm.new_buf()
            first_frame_latency = 0
            n = 0
            pcm_frame_read = 0
            read_bytes = 0
            sent_frames = 0

            audio_stream = self.elevenlabs_tts.text_to_speech_stream(msg.text)

            for chunk in self.pcm.read_pcm_stream(audio_stream, self.pcm_frame_size):
                if msg.received_ts < self.outdate_ts:
                    logger.info(f"textChan interrupt and flushing for input text: [{msg.text}], received_ts: {msg.received_ts}, outdate_ts: {self.outdate_ts}")
                    break

                if not chunk:
                    logger.info("read pcm stream EOF")
                    break

                n = len(chunk)
                read_bytes += n
                pcm_frame_read += n

                if pcm_frame_read != self.pcm.get_pcm_frame_size():
                    logger.debug(f"the number of bytes read is [{pcm_frame_read}] inconsistent with pcm frame size")
                    continue

                self.pcm.send(ten, buf)
                buf = self.pcm.new_buf()
                pcm_frame_read = 0
                sent_frames += 1

                if first_frame_latency == 0:
                    first_frame_latency = int((time.time() - start_time) * 1000)
                    logger.info(f"first frame available for text: [{msg.text}], received_ts: {msg.received_ts}, first_frame_latency: {first_frame_latency}ms")

                logger.debug(f"sending pcm data, text: [{msg.text}]")

            if pcm_frame_read > 0:
                self.pcm.send(ten, buf)
                sent_frames += 1
                logger.info(f"sending pcm remain data, text: [{msg.text}], pcm_frame_read: {pcm_frame_read}")

            finish_latency = int((time.time() - start_time) * 1000)
            logger.info(f"send pcm data finished, text: [{msg.text}], received_ts: {msg.received_ts}, read_bytes: {read_bytes}, sent_frames: {sent_frames},
                first_frame_latency: {first_frame_latency}ms, finish_latency: {finish_latency}ms"
            )
